lib/failure_detector.ml

```diff
@@ -0,0 +1,138 @@
+  t.sequence_number <- t.sequence_number + 1;
+  sequence_number
+
+let send_message client (recipient : Peer.t) message =
+  let message = Util.Encoding.pack bin_writer_message message in
+  Client.send_to client message recipient
+
+let send_ping_to client recipient = send_message client recipient Ping
+
+let send_acknowledge_to client recipient =
+  send_message client recipient Acknowledge
+
+let send_ping_request_to client recipient =
+  send_message client recipient (PingRequest recipient.socket_address)
+
+(* TODO: Totally not sure about this...
+   I don't know if I can get rid of it and only use `wait_ack_timeout` instead
+    But I feel like Lwt_condition allows ms to manage async thread and could be useful *)
+let[@warning "-32"] [@warning "-39"] rec wait_ack t sequence_number =
+  let cond =
+    find_or_add t.acknowledges sequence_number ~default:Lwt_condition.create
+  in
+  Lwt_condition.wait cond
+
+(* TODO: An async thread should be started here, with a basic sleep, but not sure at all...
+     And we should return Either Ok Timeout *)
+let wait_ack_timeout t sequence_number timeout =
+  Lwt.pick
+    [
+      (let _ = Lwt_unix.sleep (Float.of_int timeout) in
+       Lwt.return @@ Result.Error "timeout");
+      (let _ = wait_ack t sequence_number in
+       Lwt.return @@ Result.ok "received ACK");
+    ]
+
+(* Basic Knuth shuffle => https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle *)
+let knuth_shuffle known_peers =
+  let initial_array = Array.length (Array.of_list known_peers) in
+  let shuffled_array = Array.copy (Array.of_list known_peers) in
+  for i = initial_array - 1 downto 1 do
+    let k = Random.int (i + 1) in
+    let x = shuffled_array.(k) in
+    shuffled_array.(k) <- shuffled_array.(i);
+    shuffled_array.(i) <- x
+  done;
+  Array.to_list shuffled_array
+
+(* Regarding the SWIM protocol, the list of peers is not ordered.
+   Hence, I basically went for a shuffle after adding the new peer *)
+let add_peer peer_to_add t =
+  let new_peers = peer_to_add :: t.peers in
+  let shuffled_list = knuth_shuffle new_peers in
+  t.peers <- shuffled_list
+
+(* Regarding the SWIM protocol, if peer A cannot get ACK from peer B (timeout):
+   A sets B as `suspicious`
+   A randomly picks one (or several, should it also be randomly determined?) peer(s) from its list
+   and ask him/them to ping B.*)
+(* This function return the random peer, to which we will ask to ping the first peer *)
+let rec pick_random_peers (peers : Peer.t list) number_of_peers =
+  match peers with
+  | [] -> failwith "pick_random_peers"
+  | elem :: list ->
+    if number_of_peers = 1 then
+      [elem]
+    else
+      elem :: pick_random_peers list (number_of_peers - 1)
+
+(* TODO: This function will update the status of the peer to Suspicious or Faulty *)
+let update_peer _t _client _peer = failwith "undefined"
+
+(* How should I use this function? Let the client of the library defines a `peer.init` using it at `msg_handler`? *)
```

Comment:
Once this is merged and my PR is merged, I will open an issue and PR for integrating the failure detection protocol into the client, and I will handle that, so don't worry too much about this :smile:. If you would like, though, I will happily take time to show you how that will work. In fact, maybe best if we meet 1-on-1 next week to answer questions about each other's code and show examples etc.

Reply:
Got it! Removed the comment.